runtime/compiler/optimizer/InlinerTempForJ9.cpp

```diff
@@ -866,80 +831,155 @@ TR_J9InlinerPolicy::createAnchorNodesForUnsafeGetPut(TR::TreeTop* treeTop,
+      firstComparisonTree = nullComparisonTree;
+      ifTree = arrayDirectAccessTreeTop;
+      elseTree = indirectAccessTreeTop;
+      }
+   else
+      {
+      firstComparisonTree = (javaLangClass != NULL) ? lowTagCmpTree : nullComparisonTree;
+      ifTree = indirectAccessTreeTop;
+      elseTree = directAccessTreeTop;
+      }
+
+   TR::Block *joinBlock =
+         createUnsafeGetPutCallDiamond(
+               callNodeTreeTop, firstComparisonTree, ifTree, elseTree);
+
+   debugTrace(tracer(), "\t In genCodeForUnsafeGetPut, joinBlock is %d\n", joinBlock->getNumber());
+
+   TR::Block *beforeCallBlock = prevTreeTop->getEnclosingBlock();
+
+   TR::Block *arrayDirectAccessBlock = NULL;
    TR::Block *indirectAccessBlock;
-   TR::Block * directAccessBlock;
+   TR::Block *directAccessBlock;
+
    if (conversionNeeded)
       {
+      indirectAccessBlock = beforeCallBlock->getNextBlock();
+
       //Generating block for direct access
-      indirectAccessBlock = nullComparisonBlock->getNextBlock();
       directAccessBlock = TR::Block::createEmptyBlock(lowTagCmpTree->getNode(), comp(),
-      indirectAccessBlock->getFrequency());
+                                                      indirectAccessBlock->getFrequency());
       directAccessBlock->append(directAccessTreeTop);
-      directAccessBlock->append(TR::TreeTop::create(comp(),
+      directAccessBlock->append(
+            TR::TreeTop::create(comp(),
                                 TR::Node::create(directAccessTreeTop->getNode(),
-                                TR::Goto, 0, joinBlock->getEntry())));
+                                                 TR::Goto, 0, joinBlock->getEntry())));
+
+      arrayDirectAccessBlock = firstComparisonTree->getNode()->getBranchDestination()->getNode()->getBlock();
       arrayDirectAccessBlock->getExit()->insertTreeTopsAfterMe(directAccessBlock->getEntry(),
-      directAccessBlock->getExit());
+                                                               directAccessBlock->getExit());
+
       cfg->addNode(directAccessBlock);
-      cfg->addEdge(TR::CFGEdge::createEdge(directAccessBlock,  joinBlock, trMemory()));
+      cfg->addEdge(TR::CFGEdge::createEdge(directAccessBlock, joinBlock, trMemory()));
       }
    else
       {
-      directAccessBlock = nullComparisonBlock->getNextBlock();
-      indirectAccessBlock = nullComparisonTree->getNode()->getBranchDestination()->getNode()->getBlock();
+      directAccessBlock = beforeCallBlock->getNextBlock();
+
+      indirectAccessBlock = firstComparisonTree->getNode()->getBranchDestination()->getNode()->getBlock();
       indirectAccessBlock->setFrequency(VERSIONED_COLD_BLOCK_COUNT);
       indirectAccessBlock->setIsCold();
-      nullComparisonTree->getNode()->setBranchDestination(directAccessBlock->getEntry());
+
+      // If java/lang/Class is not available, the IL generated by createUnsafeGetPutCallDiamond
+      // places the direct access block on the "else" path, just to keep that nearby on the
```

Comment:
Having trouble parsing this comment in general (but it could just be me :)). Two things I'm having trouble with 
1. `on the "else" path, just to keep that nearby on the fall-through path` : I would have thought the "else path" would be something we would branch to, but maybe you are defining the "else path" in some other way that may need to be clarified.
2. "Other conditions will be placed on the fall-through" : not sure which case this is talking about, because in the int/long case, I didn't think that any other tests apart from the low tag test was to be placed on the fall-through path, i.e. even the null test was on the branch destination path out of the low tag test. 

Perhaps it would make sense to draw the order of the tests and blocks, so that we are clear on the end goal, i.e. what the IL trees will look like in the end. It feels like we will be doing some trickery (to use the phrase in the comment) and so a reader of this code in future (and indeed me, during this review) may look at the IL tree order at this point and be confused about the ordering, especially if we believe the end result ought to look quite different. This complicates the task of assessing if things are progressively coming together in a way that is moving towards the end goal we want. I'm not sure if you see what I mean, but will await your thoughts.

Reply:
> I would have thought the "else path" would be something we would branch to, but maybe you are defining the "else path" in some other way that may need to be clarified.

Right - the "else path" here is the fall-through path.  This came from terminology used in [`OMR::Block::createConditionalBlocksBeforeTree`](https://github.com/eclipse/omr/blob/4856f9da7c1f6a02e2c25308a4afcee6ceb0694d/compiler/il/OMRBlock.cpp#L1898-L1918) and in `TR_J9InlinerPolicy::addNullCheckForUnsafeGetPut` (renamed `TR_J9InlinerPolicy::createUnsafeGetPutCallDiamond` by this pull request).

I'll change references to "else path" to "fall-through path" and "if path" to "taken-side", "branch target" or similar terminology to make it more clear how these things will be placed in the IL.

> "Other conditions will be placed on the fall-through" 

For this case &mdash; the `javaLangClass == NULL`, no conversion case &mdash; at this point in the transformation, the IL looks like this.  The comparison that's performed in `block_2` is something for which we really want to branch to the "direct access" code, but it's initially set up to branch to the "indirect access" code - thus, my characterization of it as a trick.

```
block_2:
  ifacmpeq --> block_8
    aload obj
    aconst NULL
block_9:
  // direct store
  istorei <unsafe shadow sym>
    ladd
      ...
block_7:
  // join block
block_8:
  // handle indirect store processing
  goto block_7
```

After further transformations, the IL will ultimately look like this, changing the target of the first branch to the "direct access" block, with further tests inserted to branch to the direct access or indirect access block, as appropriate.

```
block_2:
  ifacmpeq --> block_9
    aload obj
    aconst NULL
block_11:
  ificmpeq --> block_9
    isArray
    iconst 1
block_10:
  iflcmpeq --> block_8
    land
      lload offs
      lconst 1
    lconst 1
block_9:
  // direct store
  istorei <unsafe shadow sym>
    ladd
      ...
block_7:
  // join block
block_8:
  // handle indirect store processing
  goto block_7
```

I'd be happy to add more comments showing these stages in the generation of the inline IL for all three cases, as I've invested the time understanding the stages of transformation of the IL.